### ISSUE TITLE
複数人での共同利用

### ISSUE THREAD

**Issue by sei0o** [sei0o/binchotan-backend#17]
ここで扱うのは、「複数の自然人が、複数のアカウントを、同じbackend経由で利用する」であって、 #10 で論じている「一人の自然人が、複数のアカウントを、同じbackend経由で利用する」ではない。

これを実装するには2つの可能性がある。すなわち、

a. 複数のアカウントを束ねる上位アカウントを作成して、それにトークン類をひも付ける。
b. Twitter Webのように、複数のアカウントを端末に対応づけて（端末を上位アカウントとみなして）管理する。

bの実装を調べる必要がある。トークン類や、上位アカウントに対応する認証情報はどこにあるのだろう？

**Comment by sei0o** [comment]
> It shouldn't be super hard to implement on a single-instance basis. Twitter works like this: You can allow other accounts to control your account under `preferences`. This allows you to post tweets etc. in Tweetdeck, which conviniently looks a lot like Mastodon.
> 
> So all it requires is that the mastodon instance stores a field in the db, saying what users are allowed to control this account. These users can then open columns in the interface which corresponds to what the interface would look like if they logged in.

https://github.com/mastodon/mastodon/issues/380 から引用。なるほどな？

**Comment by sei0o** [comment]
> It shouldn't be super hard to implement on a single-instance basis. Twitter works like this: You can allow other accounts to control your account under preferences. This allows you to post tweets etc. in Tweetdeck, which conviniently looks a lot like Mastodon.
>
> So all it requires is that the mastodon instance stores a field in the db, saying what users are allowed to control this account. These users can then open columns in the interface which corresponds to what the interface would look like if they logged in.

https://tinyurl.com/2zwx679p より引用。なるほどな？

**Comment by eniehack** [comment]
Mastodonのクライアントでbに対応しているものだと[Tooty](https://github.com/n1k0/tooty)がそうだった気がする
Pleromaだと[soapbox](https://gitlab.com/soapbox-pub/soapbox)もそう

**Comment by eniehack** [comment]
https://tabtter.jp はaの方式を採用しているっぽい
DBにアカウントの主従関係を保持していて、サービスを開くと初期化する際にそれらがbrowserに渡されるっぽい？
現在のアカウントのuser_idをlocalStorageの`_uid`に持たせて、アカウントを切り変えると値が変わっているのが確認できた

**Comment by eniehack** [comment]
aの方法だと、refresh tokenの更新をするcronが必要かも？

> How long are refresh tokens valid for?
> Refresh tokens stay valid for six months, so you will want to refresh them at least that often or more regularly. 
https://dev.to/twitterdev/handling-refresh-tokens-in-the-oauth-20-authorization-code-flow-with-pkce-with-flask-481p

**Comment by sei0o** [comment]
TootyはlocalStorageにトークンをそのまま入れている気がするなあ、XSSさえなければそれでいいのか（というより、XSSがあれば `fetch` で対象のAPIを呼び出せてしまい、cookieでも`SameSite=strict, HttpOnly`の意味がなくなってしまう？）

**Comment by sei0o** [comment]
> XSSがあれば fetch で対象のAPIを呼び出せてしまい、cookieでもSameSite=strict, HttpOnlyの意味がなくなってしまう？

いや、そんなことはないな: https://qiita.com/github0013@github/items/e57c452de28f3918bf05#comment-c3297599d3f9087c72b7

**Comment by sei0o** [comment]
考えてみた結果、aとbは便利さのベクトルが違うけれど、そのノルムはあんまり変わらないんじゃないかという気がする。backendもどうせ必要なので、aでやってみるか

**Comment by sei0o** [comment]
まずbackendでセッション（オーソドックスにcookieで）を使えるようにして、と言いたいところだがJSON-RPCでセッション管理ってどうやるんだ・・・
